### ISSUE TITLE
COPY --from doesn't work with wildcards on Windows containers

### ISSUE THREAD

**Issue by philippgille** [moby/moby#35604]
**Description**

I've got two Dockerfiles for my app - one for Linux containers and one for Windows containers. I recently changed the one for Windows containers to use multi-stage containers and to replicate the behavior of the one for Linux containers. But although the Dockerfiles are almost identical and the `COPY` line in the final multi-stage image is *exactly* the same, I get the following error message: `COPY failed: GetFileAttributesEx \\?\Volume{<some-id>}: The parameter is incorrect.`

**Steps to reproduce the issue:**

This is the Dockerfile for Linux containers,:
```Dockerfile
FROM philippgille/dotnet-libglib:2.0-sdk as builder
WORKDIR /app
COPY src/*.csproj src/
RUN dotnet restore src/hello-netcoreapp.csproj
COPY . .
RUN scripts/build.sh "fdd" "netcoreapp2.0"

FROM microsoft/dotnet:2.0-runtime
LABEL maintainer "Philipp Gille"
WORKDIR /app
COPY --from=builder /app/artifacts/hello-netcoreapp_v*_netcoreapp2.0 ./
ENTRYPOINT ["dotnet", "hello-netcoreapp.dll"]
```

It works fine when the Docker daemon is set to use Linux containers and I build the image with `docker build -f docker/Dockerfile-t local/hello-netcoreapp.`

This is the Dockerfile for Windows containers:
```Dockerfile
FROM microsoft/dotnet:2.0-sdk as builder
WORKDIR /app
COPY src/*.csproj src/
RUN dotnet restore src/hello-netcoreapp.csproj
COPY . .
RUN powershell scripts/build.ps1 -publishType "fdd" -frameworkOrRuntime "netcoreapp2.0"

FROM microsoft/dotnet:2.0-runtime
LABEL maintainer "Philipp Gille"
WORKDIR /app
COPY --from=builder /app/artifacts/hello-netcoreapp_v*_netcoreapp2.0 ./
ENTRYPOINT ["dotnet", "hello-netcoreapp.dll"]
```

I configured the Docker daemon to use Windows containers and ran `docker build -f docker/Dockerfile.nano -t local/hello-netcoreapp:nanoserver .`

**Describe the results you received:**

I get the following error message: `COPY failed: GetFileAttributesEx \\?\Volume{<some-id>}: The parameter is incorrect.`.

**Describe the results you expected:**

I expected the Docker image to be built successfully, just as the Linux container was.

**Additional information you deem important (e.g. issue happens only occasionally):**

There are two important things to note here:
1. Only the `FROM` and `RUN` in the first build stage differ from eath other and they don't have anything to do with the artifacts that are being built. I checked with a `RUN powershell ls /app/artifacts/` line in the first stage and the directory that I wanted to copy in the second stage was in fact there.
2. Proof why the `*` is what causes the failure and not any other syntactic error: When changing the `*` to `0.1.0` (which is the version number that's contained in the directory name that I want to copy), everything works fine!

I didn't find any open issue or even useful Google search result regarding this error related to Docker `COPY --from` and the [Docker COPY reference](https://docs.docker.com/engine/reference/builder/#copy) doesn't say anything about wildcard usage being operating system-dependent:
> Each <src> may contain wildcards and matching will be done using Go’s filepath.Match rules

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:09 2017
 OS/Arch:      windows/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.24)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:50:27 2017
 OS/Arch:      windows/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 11
 Running: 0
 Paused: 0
 Stopped: 11
Images: 36
Server Version: 17.09.0-ce
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd json-file logentries splunk syslog
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 16299 (16299.15.amd64fre.rs3_release.170928-1534)
Operating System: Windows 10 Pro
OSType: windows
Architecture: x86_64
CPUs: 12
Total Memory: 31.93GiB
Name: <redacted>
ID: <redacted>
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 25
 System Time: 2017-11-26T23:27:06.5705566+01:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Physical machine.

**Comment by volmedo** [comment]
I'm experiencing this very same issue. @philippgille did you find a solution (or a workaround at least) for this?

**Comment by thaJeztah** [comment]
ping @salah-khan PTAL

**Comment by volmedo** [comment]
I'm trying to make what I think is the typical use of the multistage functionality, i.e., create a final image with just the artifacts obtained in the build stage. To do so, I create a folder in the final image and throw there the main executable and a bunch of DLLs located in different folders within the project tree.

As a workaround, what I'm doing right now is preparing the artifact folder with everything needed for the application to run in the builder image by using RUN commands, which have no issues with wildcards. I then issue a single `COPY --from=builder` command in the final image to copy just this folder. Since no wildcards are involved in this COPY commands, everything works as expected.

**Comment by Ben-m-s** [comment]
@volmedo  could you please share the COPY instruction you are using to copy the folder I am trying the following and it is not working:

Step 17/19 : COPY --from=builder **/Install/Databases** /Data/
docker : COPY failed: CreateFile \\?\Volume{1ada4732-651e-11e8-b9ae-000d3af45b1c}\Install\Databases: The system cannot find the file specified.

Step 17/19 : COPY --from=builder **/Install/Databases/** /Data/
docker : COPY failed: CreateFile \\?\Volume{1ada475e-651e-11e8-b9ae-000d3af45b1c}\Install\Databases: The system cannot find the file specified.

**Comment by Ben-m-s** [comment]
I found the cause of my issue. It was a bug in my code.

**Comment by volmedo** [comment]
@Ben-m-s glad you found it!

**Comment by mpnewcomb** [comment]
Confirmed.

**Comment by awdm** [comment]
I'm experiencing the same issue. Is this considered a bug or an unsupported feature?

**Comment by nilaybhima** [comment]
I'm experiencing the same issue on a windows machine where wildcards don't work on linux containers either. Is anyone experiencing this as well? I'm using Docker Desktop for Windows V 3.1.0

**Comment by gvanbrakel** [comment]
> I'm experiencing the same issue on a windows machine where wildcards don't work on linux containers either. Is anyone experiencing this as well? I'm using Docker Desktop for Windows V 3.1.0

Yes, I do experience the same (or similar) issue, where I cannot COPY directories with wildcards.
Does not work: ``COPY  target/project-*-dependencies/*.py  /work/config/``
Does work: ``COPY  target/project-7.6-SNAPSHOT-dependencies/*.py  /work/config/``

**Comment by niemyjski** [comment]
@gvanbrakel has anyone opened a docker issue for this? I'm also seeing exactly this on windows in our own projects.

**Comment by fadamsen** [comment]
> @gvanbrakel has anyone opened a docker issue for this? I'm also seeing exactly this on windows in our own projects.

@niemyjski, this was reported as docker/for-win/issues/8333, but according to [this comment](https://github.com/docker/for-win/issues/8333#issuecomment-689032389), it's an issue in the moby build command, hence this issue.

**Comment by niemyjski** [comment]
I think we should try to reply and get more information... works everywhere but windows....

**Comment by giggio** [comment]
I can confirm the problem, but I'm getting a different error message, so I thought I'd paste it here so it can be searched.

This is the Dockerfile:

```Dockerfile
# escape=`
FROM mcr.microsoft.com/dotnet/framework/sdk:4.8-windowsservercore-ltsc2019 as choco
SHELL ["powershell.exe", "-NoLogo", "-Command"]
RUN mkdir \"${env:ProgramFiles(x86)}/c d/\"; echo b > 'C:/Program Files (x86)/c d/b.txt'; `
    mkdir \"${env:ProgramFiles(x86)}/c d/e f/\"; echo c > 'C:/Program Files (x86)/c d/e f/c.txt';

FROM mcr.microsoft.com/dotnet/framework/sdk:4.8-windowsservercore-ltsc2019
RUN mkdir \"${env:ProgramFiles(x86)}/c d/\"
COPY --from=choco [ "C:/Program Files (x86)/c d/*", "C:/Program Files (x86)/c d/" ]
```

This is the build command:

```powershell
docker build -t temp .
```

This is the error I'm getting (end of the log):
```
Step 6/6 : COPY --from=choco [ "C:/Program Files (x86)/c d/*", "C:/Program Files (x86)/c d/" ]
COPY failed: GetFileInformationByHandle \\?\Volume{4f42b98f-dd68-4bf6-8392-996da684a23d}: Incorrect function.
```

This bug is almost 5 years old...

**Comment by krthornton** [comment]
I just encountered what I believe is the same problem. Googling brought me to this issue, which doesn't appear to have had any activity since mid-2022... is this still unsolved? I am able to work around by explicitly COPYing each file I need but it would be really nice to be able to use a wildcard glob pattern instead.
